x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java

```diff
@@ -93,34 +105,126 @@ protected void readInternal(byte[] b, int offset, int length) throws IOException
+                assert streamForSequentialReads.isFullyRead() == false;
+                sequentialReadSize = NO_SEQUENTIAL_READ_OPTIMIZATION;
+                IOUtils.close(streamForSequentialReadsRef.getAndSet(null));
+            }
+        }
+
+        // read part of a blob directly; the code above falls through to this case where there is no optimization possible
         try (InputStream inputStream = blobContainer.readBlob(fileInfo.partName(part), pos, length)) {
-            int read = inputStream.read(b, offset, length);
-            assert read == length;
+            final int read = inputStream.read(b, offset, length);
+            assert read == length : read + " vs " + length;
             position += read;
         }
     }
 
+    /**
+     * If appropriate, open a new stream for sequential reading and satisfy the given read using it. Returns whether this happened or not;
+     * if it did not happen then nothing was read, and the caller should perform the read directly.
+     */
+    private boolean tryReadAndKeepStreamOpen(int part, long pos, byte[] b, int offset, int length, long currentSequentialReadSize)
+        throws IOException {
+
+        assert streamForSequentialReadsRef.get() == null : "should only be called when a new stream is needed";
+        assert currentSequentialReadSize > 0L : "should not be called if ";
+
+        final long streamLength = Math.min(currentSequentialReadSize, fileInfo.partBytes(part) - pos);
+        if (length < streamLength) {
+            // if we open a stream of length streamLength then it will not be completely consumed by this read, so it is worthwhile to open
+            // it and keep it open for future reads
+            final InputStream inputStream = blobContainer.readBlob(fileInfo.partName(part), pos, streamLength);
+            final StreamForSequentialReads newStreamForSequentialReads
+                = new StreamForSequentialReads(inputStream, part, pos, streamLength);
+            if (streamForSequentialReadsRef.compareAndSet(null, newStreamForSequentialReads) == false) {
+                // something happened concurrently, defensively stop optimizing and fall through to the unoptimized behaviour
+                this.sequentialReadSize = NO_SEQUENTIAL_READ_OPTIMIZATION;
+                inputStream.close();
+                return false;
+            }
+
+            final int read = newStreamForSequentialReads.inputStream.read(b, offset, length);
+            assert read == length : read + " vs " + length;
+            position += read;
+            newStreamForSequentialReads.pos += read;
+            assert newStreamForSequentialReads.isFullyRead() == false;
+            return true;
+        } else {
+            // streamLength <= length so this single read will consume the entire stream, so there is no need to keep hold of it, so we can
+            // tell the caller to read the data directly
```

Comment:
Should we not use this existing open stream as much as possible? We might not be able to read the full bytes from this stream, but perhaps we can use it to read everything up to streamLength, and subsequently request a new stream for the rest? This might avoid redownloading data in case where the buffer size is not a proper divisor of sequentialReadSize?

Reply:
At this point we don't have an existing open stream, we're trying to create a new one. If we can satisfy part of a read from the existing stream then we do so (see comment containing the string `the current stream didn't contain enough data for this read, so we must read more`).